packages/neos-ui-guest-frame/src/initializeGuestFrame.js

```diff
@@ -60,13 +61,48 @@ export default ({globalRegistry, store}) => function * initializeGuestFrame() {
         return;
     }
 
-    const nodes = Object.assign({}, guestFrameWindow['@Neos.Neos.Ui:Nodes'], {
-        [documentInformation.metaData.documentNode]: documentInformation.metaData.documentNodeSerialization
+    // Load legacy node data scripts from guest frame - remove with Neos 9.0
+    const legacyNodeData = guestFrameWindow['@Neos.Neos.Ui:Nodes'] || {};
+
+    // Load all nodedata for nodes in the guest frame and filter duplicates
+    const {q} = yield backend.get();
+    const nodeContextPathsInGuestFrame = findAllNodesInGuestFrame().map(node => node.getAttribute('data-__neos-node-contextpath'));
+
+    // Filter nodes that are already present in the redux store and duplicates
+    const nodesByContextPath = store.getState().cr.nodes.byContextPath;
+    const nodesAlreadyPresentInStore = {};
+    const notFullyLoadedNodeContextPaths = [...new Set(nodeContextPathsInGuestFrame)].filter((contextPath) => {
+        const node = nodesByContextPath[contextPath];
+        const nodeIsLoaded = node !== undefined && node.isFullyLoaded;
+        if (nodeIsLoaded) {
+            nodesAlreadyPresentInStore[contextPath] = node;
+            return false;
+        }
+        return true;
     });
 
+    // Load remaining list of not fully loaded nodes from the backend if there are any
+    const fullyLoadedNodesFromContent = notFullyLoadedNodeContextPaths.length > 0 ? (yield q(notFullyLoadedNodeContextPaths).get()).reduce((nodes, node) => {
+        nodes[node.contextPath] = node;
+        return nodes;
+    }, {}) : [];
```

Comment:
otherwise we modify the argument by reference and that is not the way a `reduce` should be used in js i fear? IDK

```suggestion
    const fullyLoadedNodesFromContent = notFullyLoadedNodeContextPaths.length > 0 ? (yield q(notFullyLoadedNodeContextPaths).get()).reduce((nodes, node) => ({
        ...nodes,
        [node.contextPath]: node
    }), {}) : [];
```

Reply:
My variant is faster and less memory intensive as it reuses an existing object and doesn't create new ones for each iteration. It probably make little difference here, but I try to stick with the way that's most efficient.

Reply:
Yeah, you shouldn't, as this would update the object passed to the function. But in this case a new object is created and therefore I would prefer the original implementation if we look at performance.

Reply:
But: why do we return an array instead of an object if `notFullyLoadedNodeContextPaths.length === 0`? This will result in different types...

Reply:
The return type was a mistake that Marc already fixed I think while we had a pairing session.